hashicorp/terraform
# website: Update language and CLI docs URLs

This is _most_ of the work involved in TF-294.

- [terraform-website PR](https://github.com/hashicorp/terraform-website/pull/1602)
- [v0.14 backport PR](https://github.com/hashicorp/terraform/pull/27561)

This PR rearranges the on-disk paths (and thus the URL structure) for the Terraform language docs and the Terraform CLI docs, to more or less match the logical hierarchy exposed by the sidebar navs. This includes updating any links to moved pages, in both docs and code. (Any external links to moved pages will be handled by redirects, which will be in the companion terraform-website PR.)

It also removes a few deprecated "ghost pages," which were already superseded by other pages but which we delayed deleting during the last big reorganization so we could do all the link breakage/redirections at a time of our choosing (i.e. now). 

Finally, this also removes the "this is 0.12, here's the 0.11" links from several pages in the language docs; 0.12+ is so firmly established by now that we shouldn't need these anymore.

## Notes:

- This PR will have massive failures in the broken link check CI job no matter what, because the current main branch of terraform-website lacks the necessary symlinks to even display these docs. Instead of relying on CI, I checked links manually.
- This isn't marked as 0.14 backportable, because I want to do that backport simultaneously and hand-craft it just in case something's goofy. (This touches a LOT of pages, so I'm expecting some oddities.)

## How to review this PR

This PR is huge and noisy, so reviewing it line-by-line is probably not the move. Here's my suggestions for how to review it:

- Do a local preview build, and run a link checker against it. This will confirm that I updated every link to every page that got moved. (Since the new paths don't re-use any existing paths, there's no way for an old unchanged link to be "accidentally ok.") Here's how:
    1. Download [my current link checker of choice](https://github.com/filiph/linkcheck/releases/tag/v2.0.12). Either install it to your $PATH, or just use an absolute path to the binary later.
    1. Save a text file with the following contents, to use as a "skip file":

        ```
        http://localhost:4567/(cloud|community)
        http://localhost:4567/#(cloud|cli)
        http://localhost:4567/docs/providers
        ```
    1. Clone hashicorp/terraform-website (if you don't have it already), `cd` to that directory, and fetch the upstream (or origin) remote.
    1. Run `git submodule init ext/terraform && git submodule update`.
    1. Check out the `origin/jan21_language_and_cli_urls` branch of terraform-website.
    1. `cd ext/terraform`, and check out the `origin/jan21_language_and_cli_urls` branch of terraform. Then `cd` back to the top of terraform-website.
    1. Run `make website` to get a local preview server, and wait til it starts serving content at `http://localhost:4567`.
    1. Open a new terminal window, and run `linkcheck --skip-file <THAT FILE> http://localhost:4567`
    1. When it finishes, hopefully there should be no errors or warnings, with the possible exception of some missing anchors on the front page. (Those won't appear in prod, they're just an artifact of the marketing pages being in Vercel.)
    1. Stop the preview server, go back to ext/terraform, check out the `origin/jan21_v0.14_language_and_cli_urls` branch, start the preview server again, and run the link check again. Should get the same results.
- Click around randomly in that local preview build and just make sure you're generally seeing what you expect to see.
- Spot-check individual commits. In particular:
    - Take a close look at the (very few) commits that change the navigation sidebar! Those were the template for all of the other work I did -- I did those by hand, then extracted the changes and used them to semi-automate the rest of the work, so any error there would have cascaded down.
    - Check any of the "ghost page" commits, to make sure you agree about my decision to delete those pages.
    - Check any commits that change URLs in code or otherwise outside of the docs.
- Other stuff?
    - Honestly, the safety railing of the link checker means there's only so badly I could have screwed up. But there's one main failure case it can't protect against: updating a relative link (like `./file.html`) to point to a different page with the same base filename (like something that should resolve to `/docs/language/functions/file.html` but was changed to `/docs/language/resources/provisioners/file.html` instead).

        _I'm pretty sure I prevented all of those._ But if I missed any, they'd probably have one of the following filenames:

        - list.html
        - file.html
        - configuration.html
        - syntax.html
        - show.html
        - providers.html
        - remote.html
